### PR TITLE
opencl to oneapi function port batch 1

### PR DESCRIPTION
**Pull request by pv-pterab-s**
This ports the following functions from OpenCL to OneAPI. These functions pass floating point tests on an Intel Arc 770 CPU. Certain functions such as tile, rotate, etc. do _not_ pass tests on CPU or FPGA emulation due to a reported OneAPI bug. 
 
- select
- tile
- lookup
- meanshift
- gradient
- rotate
- resize
- approx1
- approx2
- reorder
- transform

Checklist
---------

- [X] Rebased on latest master
- [X] Code compiles
- [ ] Tests pass
- [ ] Functions added to unified API
- [ ] Functions documented

### CODE REVIEW

File: src/backend/oneapi/kernel/gradient.hpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                         const int TX, const int TY, local_accessor<T> scratch)
+        : d_grad0_(d_grad0)
+        , grad0_(grad0)
+        , d_grad1_(d_grad1)
+        , grad1_(grad1)
+        , d_in_(d_in)
+        , in_(in)
+        , blocksPerMatX_(blocksPerMatX)
+        , blocksPerMatY_(blocksPerMatY)
+        , TX_(TX)
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/gradient.hpp:48]: needs to be compile time

File: src/backend/oneapi/kernel/reorder.hpp [hunk 2]
```diff
@@ -60,16 +59,17 @@ class reorderCreateKernel {
         const int xx = it.get_local_id(0) + blockIdx_x * g.get_local_range(0);
         const int yy = it.get_local_id(1) + blockIdx_y * g.get_local_range(1);
 
-        bool valid = (xx < op_.dims[0] && yy < op_.dims[1] &&
-                      oz < op_.dims[2] && ow < op_.dims[3]);
+        if (xx >= op_.dims[0] || yy >= op_.dims[1] || oz >= op_.dims[2] ||
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/reorder.hpp:62]: Revert to ```bool valid = ``` from master. Seems to be more stable (???bug)

File: src/backend/oneapi/kernel/resize.hpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    const int ix2 = (ix + 1) < in.dims[0] ? (ix + 1) : ix;
+    const int iy2 = (iy + 1) < in.dims[1] ? (iy + 1) : iy;
+
+    const VT p1 = d_in[ix + in.strides[1] * iy];
+    const VT p2 = d_in[ix + in.strides[1] * iy2];
+    const VT p3 = d_in[ix2 + in.strides[1] * iy];
+    const VT p4 = d_in[ix2 + in.strides[1] * iy2];
+
+    // d_out[ox + oy * out.strides[1]] =
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/resize.hpp:106]: stray comment

File: src/backend/oneapi/kernel/select.hpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+        const int idx0 = group_id_0 * g.get_local_range(0) + it.get_local_id(0);
+        const int idy  = group_id_1 * g.get_local_range(1) + it.get_local_id(1);
+
+        const int off = idw * oinfo_.strides[3] + idz * oinfo_.strides[2] +
+                        idy * oinfo_.strides[1];
+
+        if (idw >= oinfo_.dims[3] || idz >= oinfo_.dims[2] ||
+            idy >= oinfo_.dims[1]) {
+            return;
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/select.hpp:77]: change to ```bool valid```

File: src/backend/oneapi/kernel/select.hpp [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+            for (int idx = idx0; idx < oinfo_.dims[0];
+                 idx += g.get_local_range(0) * groups_0_) {
+                optr_pointer[idx] =
+                    (cptr[csame * idx]) ? aptr[asame * idx] : bptr[bsame * idx];
+            }
+        }
+    }
+
+   private:
+    sycl::accessor<T, 1> optr_;
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/select.hpp:106]: read/write semantics

File: src/backend/oneapi/kernel/select.hpp [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+                        idy * oinfo_.strides[1];
+
+        int ids[] = {idx0, idy, idz, idw};
+        optr_.get_pointer() += off;
+        aptr += getOffset(ainfo_.dims, ainfo_.strides, oinfo_.dims, ids);
+        cptr += getOffset(cinfo_.dims, cinfo_.strides, oinfo_.dims, ids);
+
+        if (idw >= oinfo_.dims[3] || idz >= oinfo_.dims[2] ||
+            idy >= oinfo_.dims[1]) {
+            return;
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/select.hpp:191]: bool valid

File: src/backend/oneapi/kernel/tile.hpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+        const int blockIdx_x = g.get_group_id(0) - oz * blocksPerMatX_;
+        const int blockIdx_y = g.get_group_id(1) - ow * blocksPerMatY_;
+
+        const int xx = it.get_local_id(0) + blockIdx_x * g.get_local_range(0);
+        const int yy = it.get_local_id(1) + blockIdx_y * g.get_local_range(1);
+
+        if (xx >= op_.dims[0] || yy >= op_.dims[1] || oz >= op_.dims[2] ||
+            ow >= op_.dims[3])
+            return;
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/tile.hpp:58]: bool isvalid

File: src/backend/oneapi/kernel/tile.hpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+    auto local = sycl::range(TX, TY);
+
+    int blocksPerMatX = divup(out.info.dims[0], TILEX);
+    int blocksPerMatY = divup(out.info.dims[1], TILEY);
+    auto global       = sycl::range(local[0] * blocksPerMatX * out.info.dims[2],
+                                    local[1] * blocksPerMatY * out.info.dims[3]);
+
+    getQueue().submit([&](auto &h) {
+        sycl::accessor d_out{*out.data, h, sycl::write_only, sycl::no_init};
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/tile.hpp:105]: read/write semantics

File: src/backend/oneapi/kernel/tile.hpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+#include <string>
+#include <vector>
+
+namespace arrayfire {
+namespace oneapi {
+namespace kernel {
+
+template<typename T>
+using local_accessor = sycl::accessor<T, 1, sycl::access::mode::read_write,
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/tile.hpp:25]: Test without, this should now be part of the 2020 spec as sycl::local_accessor

File: src/backend/oneapi/resize.cpp [hunk 10]
```diff
@@ -9,7 +9,7 @@
```
Comment (syurkevi) [inline src/backend/oneapi/resize.cpp:9]: This file and reorder.cpp have changed permissions to 100755. 
revert to 100644. 
and fix reorder.cpp missing (?)

Comment (syurkevi) [reply src/backend/oneapi/resize.cpp:9]: and check other files for permissions, some mismatching

**Review by syurkevi — CHANGES_REQUESTED** [review]

**Comment by syurkevi** [comment]
Also, pls check and remove any warnings the compiler will throw.

File: src/backend/oneapi/kernel/tile.hpp (hunk 9 above)
Comment (pv-pterab-s) [reply src/backend/oneapi/kernel/tile.hpp:25]: turns out intel compiler doesn't have sycl::local_accessor. we might be able to adopt this later

File: src/backend/oneapi/kernel/interp.hpp [hunk 11]
```diff
@@ -265,8 +265,8 @@ struct Interp2<Ty, Tp, 2> {
         if (method == AF_INTERP_LINEAR_COSINE ||
             method == AF_INTERP_BILINEAR_COSINE) {
             // Smooth the factional part with cosine
-            xratio = (1 - cos(xratio * af::Pi)) / 2;
-            yratio = (1 - cos(yratio * af::Pi)) / 2;
+            xratio = (1 - sycl::cospi(xratio)) / 2;
```
Comment (syurkevi) [inline src/backend/oneapi/kernel/interp.hpp:268]: :thumbsup:

**Pull request merged**